service syntax + Rust server impl

problame opened the pull request.

mbr commented on humblegen/src/ast.rs:
```diff
@@ -20,6 +25,17 @@ pub enum SpecItem {
+}
+
+impl SpecItem {
+    pub fn service_def(&self) -> Option<&ServiceDef> {
```
Docs please, even if they are simple. Currently everything else in this file is documented.

mbr commented on humblegen/src/ast.rs:
```diff
@@ -106,26 +122,88 @@ impl VariantDef {
+}
+
+#[derive(Debug)]
+pub enum ServiceRoute {
```
Do we support `DELETE` by the way? It's probably the third most common verb, followed by `PUT` then `PATCH`, in my intuition.

problame replied:
Should a `DELETE` request have a return type?
And how should non-existence of the to-be-deleted resource be represented? As a domain error (like we do for the other endpoints?)?
---

```
    http_delete ~ http_route ~ http_query? ~ "->" ~ type_ident
```
? 

Or 
```
    http_delete ~ http_route ~ http_query?
```

mbr replied:
> Should a DELETE request have a return type?

This is the same as for `POST`, possible, but not required.

Did we give this enough thought by the way? It's common (but not required) to return a 204 in these situations, which would also be valid for a `POST`. Maybe special case it for return values of `option[T]`, making `None` return an empty response?

> And how should non-existence of the to-be-deleted resource be represented? As a domain error (like we do for the other endpoints?)?

It's a domain error, since it is entirely reasonable to have both, APIs that return an error when trying to delete a non-existant entity as well as APIs that silently discard the ineffective deletion.

mbr commented on humblegen/src/ast.rs:
```diff
@@ -106,26 +122,88 @@ impl VariantDef {
+            ServiceRoute::Post { query, .. } => query,
+        }
+    }
+    pub fn rety_type(&self) -> &TypeIdent {
```
I believe we can spell out `return_type` here (and in the struct above)? No need to introduce jargon here, I think.

mbr commented on humblegen/src/elm.rs:
```diff
@@ -180,6 +181,7 @@ fn render_type_ident(type_ident: &ast::TypeIdent) -> String {
         ast::TypeIdent::BuiltIn(atom) => render_atom(atom),
         ast::TypeIdent::List(inner) => format!("List {}", opt_parens(render_type_ident(inner))),
         ast::TypeIdent::Option(inner) => format!("Maybe {}", opt_parens(render_type_ident(inner))),
+        ast::TypeIdent::Result(_okk, _err) => todo!(),
```
I think this could straightforwardly be added as `format!("Result {} {}", opt_parens(render_type_ident(err)), opt_parens(render_type_ident(ok)))` --- note the switched type order.

But it's fine to leave that for the Elm implementor.

mbr commented on humblegen/src/humble.pest:
```diff
@@ -18,22 +19,47 @@ doc_comment = { doc_comment_line+ }
+service_definition = { doc_comment? ~ "service" ~ camel_case_ident ~ service_def }
+http_route = ${http_route_segment+ }
+http_route_segment = ${
+    "/" ~ (kebab_case_ident|http_route_segment_arg) // s/snake_case_ident/kebab_kase/
```
`kebab-case` (in comment)

mbr commented on humblegen/src/parser.rs:
```diff
@@ -1,21 +1,32 @@
+    // AST transformations
+    embeds::resolve_embeds(&mut ast);
+
+    // TODO: type checks
```
Let's open issues instead of peppering our code with `TODO`s (of which I am the most guilty myself).

mbr commented on humblegen/src/parser/embeds.rs:
```diff
@@ -0,0 +1,89 @@
+use crate::ast::*;
```
`*` import again. I believe the other code uses prefixed? I find that a lot more tidy.

Also, a few sentences describing the transform would be nice.

problame replied:
```
* import again. I believe the other code uses prefixed? I find that a lot more tidy.
```
Nope, parent mod `parser` does the same thing, and I think that's totally fine given that the parser's job is to construct the AST.

mbr commented on humblegen/src/parser/embeds.rs:
```diff
@@ -0,0 +1,89 @@
+fn spec_resolve_embeds_one_level(spec: &mut Spec) -> bool {
+    let mut changed = false;
+
+    // FIXME: ast should support visitor pattern
```
Issue here as well.

mbr commented on humblegen/src/rust.rs:
```diff
@@ -150,9 +162,389 @@ fn render_atom(atom: &ast::AtomType) -> TokenStream {
+            services: Vec<Service>,
+        }
+
+        use std::net::SocketAddr;
```
Move this up?

mbr commented on humblegen/src/rust.rs:
```diff
@@ -150,9 +162,389 @@ fn render_atom(atom: &ast::AtomType) -> TokenStream {
+    }
+}
+
+struct ServiceRoute {
```
Do all the structs we used here not allow derivation of `Debug`? I would want that on every `struct` where it is free.

problame replied:
not quite free, but did it now anyways

mbr commented on humblegen/tests/rust/service/spec.humble:
```diff
@@ -0,0 +1,83 @@
+    .. MonsterData3,
+}
+
+struct MonsterData3 {
```
At some point, we should separate the "educational" specs from those testing the parser/generator, as this is mighty confusing...

mbr commented on humblegen/tests/rust/service/spec.humble:
```diff
@@ -0,0 +1,83 @@
+    max_age: option[i32],
+}
+
+/// WHY?
```
I don't know =(

problame commented:
Quick notes as I'm in a hurry, will comment more this afternoon:

* I wouldn't merge the branch as is if the `async-trait` patch is a show-stopper for you. Upstream doesn't react to [our PR](https://github.com/dtolnay/async-trait/pull/96) and the discussions in the corresponding upstream issue make the impression that they'd rather wait out changes in the larger async + hyper ecosystem then merging it.
  * I do not see a way to have async handler impls without our patch, feel free to investigate further.
  * We don't want users to start implementing async handlers and later tell them they need to convert their async impls to sync code
* `However this is underdocumented in many places, which makes it hard for outsiders (like myself ;)) to quickly grok what's going on.`
  * Could you please be more specific about this? Do you need a block-comment somewhere that explains how the server is put together? I do not believe that function-level comments in the parser etc. are useful, but sure, I can add those, too.
* `DELETE` verb mentioned in a comment should become an issue if we want to merge quickly
* Most importantly, the review doesn't answer the question whether the _functionality_ it provides _today_ is sufficient and working for you. Have you written a practical API spec and played with the generated code?

mbr commented:
>     * I wouldn't merge the branch as is if the `async-trait` patch is a show-stopper for you. Upstream doesn't react to [our PR](https://github.com/dtolnay/async-trait/pull/96) and the discussions in the corresponding upstream issue make the impression that they'd rather wait out changes in the larger async + hyper ecosystem then merging it.

I can live with it, I think. I will have to look at it in Detail for a workaround.
Could we vendor/reexport `async_trait` in our runtime?

>       * I do not see a way to have async handler impls without our patch, feel free to investigate further.
>       * We don't want users to start implementing async handlers and later tell them they need to convert their async impls to sync code

That's the thing, the "high performance" everyone is clamoring for is served by `async` usually, but there'd be no shame in just having a thread pool processing synchronous requests, making the hyper-async part just play the role of reverse proxy if we only expect DB operations and such.

That does not hold true though if network IO is involved (i.e. making requests to other services), as then handling a request cannot be expected to only block on the single thing it would block on anyway (DB IO).

>     * `However this is underdocumented in many places, which makes it hard for outsiders (like myself ;)) to quickly grok what's going on.`
>       
>       * Could you please be more specific about this? Do you need a block-comment somewhere that explains how the server is put together? I do not believe that function-level comments in the parser etc. are useful, but sure, I can add those, too.

Personally I find them very helpful, but I am willing to be convinced otherwise. It might just be me though, I used full comments usually as a marker that I have revisited and refactored that function sufficiently =).

>     * `DELETE` verb mentioned in a comment should become an issue if we want to merge quickly

I'd settle for an `unimplemeted!`, so it can be used in specs and possibly other backends. Deleting using `POST` somehow feels like bad form.

>     * Most importantly, the review doesn't answer the question whether the _functionality_ it provides _today_ is sufficient and working for you. Have you written a practical API spec and played with the generated code?

Guilty as charged, working on it.

problame commented on humblegen-rt/src/regexset_map.rs:
```diff
@@ -0,0 +1,75 @@
+//!
+//! The `GetResult` contains a reference to the matching entry.
+
+/// Refer to module-level docs.
```
I think the module-level comment is a better place to explain the role of an `Entry` in a `RegexSetMap`. And it already does that pretty well IMHO.

problame commented on example/src/main.rs:
```diff
@@ -0,0 +1,31 @@
+
+#[tokio::main]
+async fn main() {
+    serve_and_run().await;
```
use of the `tokio::main` proc-macro results in pretty useless error messages if there is any error in the `main()`, but sure

problame commented on humblegen-rt/src/handler.rs:
```diff
@@ -0,0 +1,28 @@
+impl ServiceError {
+    pub(crate) fn to_hyper(&self) -> Response<Body> {
+        let (code, msg) = match self {
+            ServiceError::Authentication => (401, "authentication failure".to_string()),
```
Error responses in general have been unified in a commit I just pushed.

problame commented:
I pushed a bunch of commits that address all the review comments that didn't need further discussion.

* Code comments should be more numerous and maybe useful now
* There's `./humblespec` that explains some concepts of the humblespec format (only about the`service` directive ATM).

A few of the review comments still need discussion.

---

**Responses to your last comment:**

> Could we vendor/reexport async_trait in our runtime?

No, because it's a proc macro. We could ship our own wrapper-proc-macro though, but that'd mean we'd have `humblegen-rt` _and_ `humblegen-rt-macros`.

> That's the thing, the "high performance" everyone is clamoring for is served by async usually, but there'd be no shame in just having a thread pool processing synchronous requests, making the hyper-async part just play the role of reverse proxy if we only expect DB operations and such.

> That does not hold true though if network IO is involved (i.e. making requests to other services), as then handling a request cannot be expected to only block on the single thing it would block on anyway (DB IO).

Don't know what to make of this...


>     * `However this is underdocumented in many places, which makes it hard for outsiders (like myself ;)) to quickly grok what's going on.`
>       
>       * Could you please be more specific about this? Do you need a block-comment somewhere that explains how the server is put together? I do not believe that function-level comments in the parser etc. are useful, but sure, I can add those, too.
> 
> Personally I find them very helpful, but I am willing to be convinced otherwise. It might just be me though, I used full comments usually as a marker that I have revisited and refactored that function sufficiently =).
> 

I committed a bunch of overview documentation in `./humblespec` and attempted to improve the Rust docs of the generated code (semi-successful though, and we need to find a way to render the example project's docs on docs.rs).

---

problame commented:
One more thing wrt CI:

We want to run build checks on the example project in CI, but GitHub Actions [doesn't allow changing the `working-directory` for actions](https://github.community/t5/GitHub-Actions/Can-t-specify-quot-working-directory-quot-for-actions-Only-for/td-p/41992).

The GitHub actions `cargo` action that we use ATM also doesn't support switching the working directory.

How do we solve this? Bash script and don't use the `cargo` action?

problame commented:
I think I have addressed alsmost all review comments now:

* [x] our `async-trait` is now tracked as a git submodule, crate has different name => users don't need to use a `[patch]` stmt in Cargo.toml
* [x] converted CI to gitlab, dropped github actions
* [ ] RegexSetMaps docs - I think what we have now is sufficient
* [ ] `DELETE` support
    * [x] as dicussed on Zoom, implemented with the same signature as `GET`
    * [x] rust server impl
    * [x] added a `()` empty type for use as `result[()][DomainError]` (elm unimplemented)
    * [ ] no special-case handling to produce a 204 in case the delete succeeded
        * I think this PR is the wrong time to find the optimal way to allow a handler to override the 200 that we usually return for `Ok()` handler responses
* [x] prettyfied commit history, the last patch is the actualy feature, all others are independent (but preparatory) features
